### ISSUE TITLE
DZ pyramid builder problem

### ISSUE THREAD

**Issue by alonit** [jcupitt/libvips#527]
Hi,

I am using the C API VIPS 8.3.3 64bit version over Windows 7 64bit, 12GB RAM. and Intel 4 cores CPU HyperThreaded.

I am scanning 1508 (grid size: 26 by 58) images where each image is 2356 by 1772 pixels , 3 bands, sRGB and save them as jpeg files under a folder as follows: 1.jpg, 2.jpg, ... , 1508.jpg.

I want to build a DZ pyramid for presenting those images in a viewer as a whole.

I use vips_arrayjoin which output a single stitched image with size 61256 by 102776 pixels, 3 bands, sRGB. When I send this big image to vips_dzsave I get an exception in the middle of its execution: "Exception thrown at 0x0000000063A47CB3 (libgobject-2.0-0.dll) in Asi.Bsw.VIPS.Client.exe: 0xC0000005: Access violation reading location 0x0000000000000008." from the run time.

Is there anyway to overcome this problem ?

The code looks like:

`
//1) Stitch the whole image out of 1508 images:

//1508 = 26x58 grid size
    VipsImage\* images[1508]; 

```
int imageIndex = 0;

for (int i = 1; i < 59; i++)
{
    for (int j = 1; j <= 26; j++)
    {
        imageIndex = (26 * (i - 1)) + j;

        VipsImage *Im1;

        char inputname[256];
        sprintf(inputname, "%s%d.jpg", "C:\\VIPS_Imaging\\Colon\\", imageIndex); 

        //get next image in grid
        if (!(Im1 = vips_image_new_from_file(inputname,
            "access", VIPS_ACCESS_SEQUENTIAL_UNBUFFERED,
            NULL)))
        {
            vips_error_exit(NULL);
        }

        //Save it to images array.
        images[imageIndex-1] = Im1;
    }
}

//Whole image
VipsImage *ImOutTotal = NULL;

//Join array of images
if (vips_arrayjoin(images,
    &ImOutTotal,
    1508,
    "across", 26,
    NULL))
{
    const char* cError = vips_error_buffer();

    vips_error_exit(cError);
}
```

//2) Build the DZ pyramid fro the ImOutTotal (Whole image):

double background[3] = { 1,1,1 };

```
if (vips_dzsave(ImOutTotal, "C:\\VIPS_Imaging\\Colon\\My_Tiles",
    "layout", VIPS_FOREIGN_DZ_LAYOUT_DZ,

    "suffix", ".jpg[Q=75]", //(1 - 100). Default 75.

    "overlap", 1,

    "tile_size", 254,

    "background", background,

    "depth", VIPS_FOREIGN_DZ_DEPTH_ONETILE,

    "centre", false, 

    "angle", VIPS_ANGLE_D0, 

    "container", VIPS_FOREIGN_DZ_CONTAINER_FS, 

    "properties", false,

    //"compression" , 1,
    NULL))   //Here the run-time throw the exception!!!
{
    const char* cError = vips_error_buffer();

    vips_error_exit(cError);
}
```

`

Thanks.

**Comment by jcupitt** [comment]
You're not passing background correctly, try removing that line. It really does have to be a `VipsArrayDouble` object, it's a bit awkward. I don't think you need to specify the background anyway.

**Comment by alonit** [comment]
I removed the "background" line and it still crashes with: 
Exception thrown at 0x0000000063A47CB3 (libgobject-2.0-0.dll) in Asi.Bsw.VIPS.Client.exe: 0xC0000005: Access violation reading location 0x0000000000000008.

VIP console message when crash occure:

** (Asi.Bsw.VIPS.Client.exe:4932): ←[1;35mCRITICAL←[0m **: gsf_output_write: ass
ertion 'output != NULL' failed

** (Asi.Bsw.VIPS.Client.exe:4932): ←[1;35mCRITICAL←[0m **: gsf_output_close: ass
ertion 'GSF_IS_OUTPUT (output)' failed

** (Asi.Bsw.VIPS.Client.exe:4932): ←[1;35mCRITICAL←[0m **: gsf_output_set_error:
 assertion 'GSF_IS_OUTPUT (output)' failed

**Comment by jcupitt** [comment]
Hmm strange. Did you remove the old `My_Tiles` directory before running the program? What if you run `dzsave` at the command-line, does that work? 

```
vips.exe dzsave somefile.jpg x
```

Try removing all of the options to dzsave and see if it works.

**Comment by alonit** [comment]
I removed My_Tiles directory and all dzsave options, but it is still crashes.

**Comment by jcupitt** [comment]
It sounds like it's unable to create a subdirectory for some reason. 

Try writing your huge image to a tif file, then running dzsave from that at the command-line.

I would also try zip output.

**Comment by alonit** [comment]
When running, it is creating the subdirectory stared at '17' down to '11'  with the jpeg tiles inside, and then crashes.

I'll try saving tif file before dzsave it - but it will not perform as good as passing it directly to `vips_dzsave`: Write huge tiff to disc, and then read it for `vips_dzsave`

Can I call vips_dzsave (C API) with optional arguments to DZ save `ImOutTotal` (the huge image) using .tif suffix and compressed with jpeg with quality-factor X (for example X=90).

**Comment by alonit** [comment]
I tried it by calling:

```
vips_dzsave(ImOutTotal, "C:\\VIPS_Imaging\\Colon\\My_Tiles",
"suffix",".tif[compression=jpeg,Q=90]",
NULL);
```

but it create `My_Tiles-B1TINY` folder and then blocked inside `vips_dzsave` implementaion. (CPU usage is 0-1%).

**Comment by jcupitt** [comment]
This is very strange, loads of people are using that binary without problems, as far as I know.

Could you try the command-line interface and zip output?

**Comment by alonit** [comment]
Here are the results:







Input:
 ImageJoined494.jpg and ImOutTotal_494 are single image joined from 494 .jpg images of 2356x1772 pixels each, 3 bands.

Tests:
1. Command line: vips.exe dzsave ImageJoined494.jpg x 
**- succeeded**

2. C API: vips_dzsave(ImOutTotal_494, "C:\\VIPS_Imaging\\Colon\\My_Tiles", NULL) 
**- succeeded**

So VIPS with .jpg works for me.

3. C API: vips_image_write_to_file(ImOutTotal_494, "G:\\Colon\\ImageJoined494.tif", NULL)
- The ImageJoined494.tif file is created under G:\\Colon\\ but when it reaches 4GB it became invalid - no properties are available on it anymore so I think I need help on how to generate big .tif images in VIPS. 

In addition the CPU is busy (15%) inside vips_image_write_to_file for about 10 minutes without return.

Next test I tried (and succeeded) to create smaller .tif  file joined from just 78 images:
4. C API: vips_image_write_to_file(ImOutTotal_78, "G:\\Colon\\ImageJoined78.tif", NULL) 
**- succeeded** - So it seems again that the problem is to create huge .tif files > 4GB

5. C API: vips_image_write_to_file(ImOutTotal, "G:\\Colon\\ImageJoined78.tif", 
		"compression", VIPS_FOREIGN_TIFF_COMPRESSION_JPEG,
		"tile", TRUE, 
		NULL)
- **- succeeded**

6. Command line: vips.exe dzsave C:\VIPS_Imaging\Colon\ImageJoined494.jpg y --suffix .tif 
- **failed!!!** immediately with the following messages:
VipsForeignSave: ".tif" is not a known buffer format
VipsForeignSave: ".tif" is not a known buffer format
VipsForeignSave: ".tif" is not a known buffer format
VipsForeignSave: ".tif" is not a known buffer format
VipsForeignSave: ".tif" is not a known buffer format
VipsForeignSave: ".tif" is not a known buffer format
VipsForeignSave: ".tif" is not a known buffer format
VipsForeignSave: ".tif" is not a known buffer format
wbuffer_write: write failed

7. 6. Command line: vips.exe dzsave C:\VIPS_Imaging\Colon\ImageJoined78.jpg y --suffix .tif 
- **failed!!!** immediately with the following messages:
VipsForeignSave: ".tif" is not a known buffer format
VipsForeignSave: ".tif" is not a known buffer format
VipsForeignSave: ".tif" is not a known buffer format
VipsForeignSave: ".tif" is not a known buffer format
VipsForeignSave: ".tif" is not a known buffer format
VipsForeignSave: ".tif" is not a known buffer format
VipsForeignSave: ".tif" is not a known buffer format
VipsForeignSave: ".tif" is not a known buffer format
wbuffer_write: write failed

8. C API: vips_dzsave(ImOutTotal_494, "suffix", ".tif[compression=jpeg,Q=90]", "C:\\VIPS_Imaging\\Colon\\My_Tiles", NULL) 
- failed. Create `My_Tiles_TIF-L8Q0NY` folder is empty. vips_dzsave not returned and CPU usage is 0%
8. C API: vips_dzsave(ImOutTotal_78, "suffix", ".tif[compression=jpeg,Q=90]", "C:\\VIPS_Imaging\\Colon\\My_Tiles", NULL) 
- failed. Create `My_Tiles_TIF-L8Q0NY` folder is empty. vips_dzsave not returned and CPU usage is 0%

I think i have problem generating big .tif images with VIPS and also DZ save into .tif suffix files from CL or C API

**Comment by alonit** [comment]
I also tried:

	vips_image_write_to_file(ImOutTotal_494, "G:\\Colon\\ImageJoined494.dz", 
		"compression", VIPS_FOREIGN_TIFF_COMPRESSION_JPEG,
		"tile", TRUE, 
		NULL)

but it return with cError:	0x0000000069fcc780 "dzsave: no property named `compression'\n"

**Comment by jcupitt** [comment]
Ooops, sorry, I forgot that dzsave uses the buffer output functions, so it only supports jpg, png, rad and webp tiles. 

You turn on bigtiff output with:

```
    "bigtiff", TRUE,
```

in the option list.

**Comment by alonit** [comment]
Hi John,

Thank you for your replies.

ImOutTotal_1508 is `VipsImage `buffer holding 1508 stitched images (2356x1772 pixels each, 3 bands, 24bpp).

I'v tried to save ImOutTotal_1508 to disc and set `"bigtiff", TRUE,` but I still got exceptions while calling `vips_image_write_to_file `with the following parameters: 

1 . Write to .tif file with jpeg compression:

```
vips_image_write_to_file(ImOutTotal_1508, "G:\\Colon\\ImageJoined1508.tif", 
"compression", VIPS_FOREIGN_TIFF_COMPRESSION_JPEG, 
"tile", TRUE, 
bigtiff", TRUE,
NULL)
```
-> Exception thrown at 0x0000000000000000 in Asi.Bsw.VIPS.Client.exe: 0xC0000005: Access violation executing location 0x0000000000000000.

2. Write to .tif file without jpeg compression:
```
vips_image_write_to_file(ImOutTotal_1508, "G:\\Colon\\ImageJoined1508.tif", 
"tile", TRUE, 
bigtiff", TRUE,
NULL)
```
-> Exception thrown at 0x0000000000000000 in Asi.Bsw.VIPS.Client.exe: 0xC0000005: Access violation executing location 0x0000000000000000.

1. Is there mistakes in my implementation?
2. Generally, I'm trying to build pyramid out of given 1508 jpeg images, 2356x1772 pixels each, how can I create this pyramid using the C API with emphasis on performance?

Thanks a lot in advanced

**Comment by jcupitt** [comment]
Does it crash some way into writing the bigtiff, or is it immediate? If it makes part of the output, how big is the bit it makes?

Could you try at the command-line?

```
vips.exe tiffsave somefile.jpg x.tif --bigtiff
vips.exe tiffsave somefile.jpg x.tif --bigtiff --tile
vips.exe tiffsave somefile.jpg x.tif --bigtiff --tile --compression=jpeg
```

Do they all work?

Just save to a `.dz` file, it should be quick. Use zip output, it's much faster on Windows.

**Comment by alonit** [comment]
Now, after restarting my machine, and call 

```
vips_image_write_to_file(ImOutTotal_1508, "G:\\Colon\\ImageJoined1508.tif", 
"tile", TRUE, 
bigtiff", TRUE,
NULL)

```

I get the following error:

`cError	0x0000000069fcc780 "vips__file_open_read: unable to open file \"C:\\VIPS_Imaging\\Colon\\510.jpg\" for reading\nwindows error: The program issued a command but the command length is incorrect.\r\n\nvips__region_start: start funct...	const char *`


510.jpg if the 510 jpg file stitched into `ImOutTotal_1508` by calling:

```
vips_arrayjoin(images,
		&ImOutTotal,
		IMAGES,
		"across", 26,
		NULL)
```

**Comment by alonit** [comment]
I also tested the C API with 494 (grid: 26 by 19) stitched images and it works for me:

vips_image_write_to_file(ImOutTotal_494, "G:\\Colon\\ImageJoined494.tif", 
bigtiff", TRUE,
NULL)

vips_image_write_to_file(ImOutTotal_494, "G:\\Colon\\ImageJoined494.tif", 
"tile", TRUE, 
bigtiff", TRUE,
NULL)

vips_image_write_to_file(ImOutTotal_494, "G:\\Colon\\ImageJoined494.tif", 
"compression", VIPS_FOREIGN_TIFF_COMPRESSION_JPEG,
"tile", TRUE, 
bigtiff", TRUE,
NULL)

but when I goes up to 520 stitched images (grid: 26 by 20) it fails internally at `vips_image_write_to_file ` when accessing 510.jpg with error:

`vips__file_open_read: unable to open file "C:\VIPS_Imaging\Colon\510.jpg" for reading windows error: The program issued a command but the command length is incorrect. vips__region_start: start function failed for image C:\VIPS_Imaging\Colon\510.jpg`

**Comment by jcupitt** [comment]
Oh interesting! Perhaps this is a Windows 7 limit and you can't have more than 500 files open at once?

Your options would seem to be:

* Update to win10. 
* Run `dzsave` in a virtual machine under linux. You can make this pretty seamless: you can use docker to trigger the `dzsave` from Windows `cmd.exe`, and block until it finishes. Performance will be better than under Windows as well. 
* Do the join in two stages: do the top half and arrayjoin to a bigtiff, do the bottom half to another temporary file, then join the two halves to `dzsave`.

**Comment by jcupitt** [comment]
I found this:

http://stackoverflow.com/questions/870173/is-there-a-limit-on-number-of-open-files-in-windows/870224#870224

There's a maximum of 512 files open at once on Windows. It looks like we could raise this to 2,048 with some changes to vips, but it can't easily go any higher than that.

**Comment by jcupitt** [comment]
I added a call to `_setmaxstdio()` to `vips_init()`, this will be in vips 8.4, due in a few days.

**Comment by jcupitt** [comment]
Here's a test build of 8.4 incorporating this change:

http://www.vips.ecs.soton.ac.uk/development/8.4/vips-dev-w64-all-8.4.0c.zip

If you have time to test it, that'd be great!

**Comment by jcupitt** [comment]
This summarizes the 8.3 -> 8.4 changes:

https://github.com/jcupitt/libvips/blob/master/whatsnew-8.4.md

I'm not sure any of them will affect what you are doing.

**Comment by alonit** [comment]
I tested 8.4 and it works very well! - I can now save single image (joined from 1508 images) to a .tif file on disc with jpeg compression.

As I understand, the VipsImage buffer output (holding 1508 joined images) cannot be used by C API `vips_dzsave ` function to save .tif tiles, as `vips_dzsave ` is support only jpg, png, rad and webp tiles - Is this correct?  Is version 8.4 support additional possibilities?

I can't use C API `vips_dzsave ` to create .jpg tiles because it is limited to max of 65Kx65K image.
I also can't use .tif tiles - as you explained above, this is not supported by C API `vips_dzsave ` .
So, I tried `vips_dzsave ` with .png tiles and it turn out to be very slow on my Intel i7/12GB RAM/SSD system:

```
vips_dzsave(ImOutTotal_1508, "G:\\Colon2\\My_Tiles_TIF",
	"suffix", ".png",
       NULL)
``` -> Pyramid build is slow using this option.

Given that I have access only to VIPS C API, how can I use the output buffer to create well performed pyramid builder (for example the output buffer size is 70Kx100K)?

Many thanks

**Comment by jcupitt** [comment]
That's great! Thanks for testing it.

You can use dzsave to write images with jpg tiles. You can't have a single jpg image of more than 65536 x 65536 pixels, but it's fine to have loads of small jpg tiles.

You can't have tif tiles in `dzsave`, but no deepzoom reader will let you read tif tiles (as far as I know), so I don't think this is a problem. 

Yes, png is a very, very slow format.

For best performance on Windows, use jpg tiles and a zip container. The zip container is very important, it makes `dzsave` maybe 3 or 4 times faster, and the pyramid is much easier to manage as a zip file rather than a huge directory.

**Comment by alonit** [comment]
Thank you for you replies and great support, I really appreciate it.

Using v8.3 a call to `_setmaxstdio(2048)` does not help using more than 512 file in my process. I validated this set with a call to _getmaxstdio() that return 2048 just before using VIPS C API, but while executing VIPS I got:
`vips__file_open_read: unable to open file "C:\VIPS_Imaging\Colon\510.jpg" for reading windows error: The program issued a command but the command length is incorrect. vips__region_start: start function failed for image C:\VIPS_Imaging\Colon\510.jpg`

With v8.4 I was able to create pyramids out of a single 1508 joined image and save its tiles as jpg and as zipped jpg, but from time to time I get an error while vips_dzsave is in the middle of its execution:

The call is:

vips_dzsave(ImOutTotal_1508, "G:\\Colon2\\My_Tiles_ZIP2",
    "suffix", ".jpg[Q=75]",
     "container", VIPS_FOREIGN_DZ_CONTAINER_ZIP, 
       NULL)

and the error is:
`0x000000006a01cac0 "VipsJpeg: out of order read at line 192\nVipsJpeg: out of order read at line 192\n"`

Thanks

**Comment by jcupitt** [comment]
For the out of order error, swap `VIPS_ACCESS_SEQUENTIAL_UNBUFFERED` for `VIPS_ACCESS_SEQUENTIAL`. 

The unbuffered version keeps just a single scanline for each input file, and for complex pipelines it can sometimes not be enough if threads get out of order. The plain seq version keeps a few 100 scanlines, so memory use will be higher, but there's some margin for error.

**Comment by alonit** [comment]
Test 1:
I called `vips_concurrency_set(1)` and it works fine with `VIPS_ACCESS_SEQUENTIAL_UNBUFFERED` , low memory footprint and reasonable time.

Test 2:
Without calling `vips_concurrency_set(1)` , It works fine with `VIPS_ACCESS_SEQUENTIAL `but memory usage was huge (~15GB peak), and very slow in time (x4 in my test compared to Test 1), So i think to use VIPS with `vips_concurrency_set`(1) and open new thread in my user level for each pyramid build transaction, and get transactions parallelism. Is this kind of parallelism can be supported by VIPS ( can VIPS engine run in multi-threaded environment, from multiple threads concurrently)?

Many thanks

**Comment by jcupitt** [comment]
Yes, vips is thread-safe, you can build many pyramids at once. You can pass `VipsImage*` between threads too, and read the same image from many threads at once. Of course you may start to saturate your disc drive. 

I see a speedup with many threads, though it's not huge on my 2 core, 4 thread laptop. For a 20,000 x 20,000 RGB jpg image I see:

```
$ time vips dzsave wtc2.jpg x.zip --vips-concurrency=4
real	0m10.014s
user	0m17.780s
sys	0m6.412s
$ time vips dzsave wtc2.jpg x.zip --vips-concurrency=1
real	0m11.415s
user	0m12.700s
sys	0m1.056s
```

I'll try on a bigger desktop machine.

**Comment by alonit** [comment]
OK Thanks,  I'll check this again: `vips-concurrency=1 and VIPS_ACCESS_SEQUENTIAL_UNBUFFERED` VS `vips-concurrency=4 and VIPS_ACCESS_SEQUENTIAL` .

Using VIPS C API:
1. Can the C API provide function calls performance statistics such as cpu usage/memory/time?
2. Can the C API provide progress statistics while calling its functions?
3. Can the C API provide image properties given valid `VipsImage*` ?

Thanks you very much for the extra help.

**Comment by jcupitt** [comment]
1. I use `time` to get an overview of a whole pipeline, or `callgrind` to profile individual functions. libvips has a very simple built-in progress reporter: use [`vips_progress_set()`](http://www.vips.ecs.soton.ac.uk/supported/current/doc/html/libvips/VipsImage.html#vips-progress-set) to turn it on, or the `--vips-progress` command-line flag. It gives some timing statistics.

2. You can add your own progress feedback. Call [`vips_image_set_progress()`](http://www.vips.ecs.soton.ac.uk/supported/current/doc/html/libvips/VipsImage.html#vips-image-set-progress) on an image, then listen for the start / eval / stop signals on the image. It tracks stats like number of pixels processed, percent complete, and estimated time to completion.

3. Sure, just read the properties out of the image.

**Comment by jcupitt** [comment]
Oh, there's a leak tracker too, call `vips_leak_set()` and you'll get a list of leaked objects on exit (if any), plus a measure of memory high water. 

There's also a detailed profiler:

http://libvips.blogspot.co.uk/2013/11/profiling-libvips.html

It can track things like time spent waiting on thread synchronisation, or memory cycling.

**Comment by jcupitt** [comment]
I tried on a big desktop machine (6 cores, 12 threads):

```
$ time vips dzsave wtc2.jpg x 
real	0m5.841s
user	0m16.300s
sys	0m9.696s
$ time vips dzsave wtc2.jpg x --vips-concurrency=1
real	0m12.769s
user	0m13.652s
sys	0m1.628s
```

Again, that's a 20k x 20k pixel rgb jpg image.

**Comment by jcupitt** [comment]
I tried doing it via `arrayjoin` as well using Python:

```python
 #!/usr/bin/env python

import sys

import gi
gi.require_version('Vips', '8.0')
from gi.repository import Vips

Vips.leak_set(True)
Vips.progress_set(True)

images = []
for name in sys.argv[2:]:
    im = Vips.Image.new_from_file(name,
                                  access = Vips.Access.SEQUENTIAL_UNBUFFERED)
    images.append(im)

join = Vips.Image.arrayjoin(images, across = 26)
```

I tested like this:

```
$ vips crop wtc.jpg x.jpg 2000 2000 2356 1772
$ mkdir test
$ cd test
$ for i in {1..1508}; do cp ../x.jpg $i.jpg; done
$ ulimit -n 10000
$ ../arrayjoin.py x.dz *.jpg
arrayjoin.py temp-3017: 61256 x 102776 pixels, 12 threads, 61256 x 16 tiles, 384 lines in buffer
arrayjoin.py temp-3017: done in 112s          
memory: high-water mark 9.79 GB
```

I added `Vips.concurrency_set(1)` to the program and ran it again:

```
$ ./arrayjoin.py x.dz *.jpg
arrayjoin.py temp-3017: 61256 x 102776 pixels, 1 threads, 61256 x 16 tiles, 256 lines in buffer
arrayjoin.py temp-3017: done in 191s          
memory: high-water mark 1.01 GB
```

On Windows you'll need to use zip output or it'll run rather slowly, something like:

```
$ ./arrayjoin.py x.dz[container=zip] *.jpg
```

**Comment by alonit** [comment]
I tried to register to the progress notifications as follows:

define the callback:

```
//The ::eval signal is emitted once per work unit (typically a 128 x 128 area of pixels) during image //computation.
void ASI_user_functionEval(VipsImage *image,
	gpointer   progress,
	gpointer   user_data)
{
//Handle notification here...
}
```

register the callback:

```
g_signal_connect(inputImage, "eval", G_CALLBACK(ASI_user_functionEval), NULL);
vips_image_set_progress((VipsImage*)inputImage, true);
```

and then call `vips_dzsave(inputImage,...);`

but the callback is not called during vips_dzsave execution.

Did i missed something?

Thanks

**Comment by alonit** [comment]
As you can see above, I'm calling `vips_image_set_progress` just after `g_signal_connect(...)`, SO I will try to call it before it.

Thanks.

**Comment by jcupitt** [comment]
You need to call before you use the image as well. The sequence should be:

1. `vips_arrayjoin(a, &x, NULL)`
2. `vips_image_set_progress(x, TRUE)`
3. `g_signal_connect(x, ...)`
4. `vips_dzsave(x, ...)`

**Comment by jcupitt** [comment]
Oh heh well done, I hadn't thought of that. Yes, you'll need to remove the `vips_progress_set(true);`.

No, you can only have one set of signals triggered per evaluation loop, since only one image is really being evaluated. The internal vips one you get with `vips_progress_set(true);` is very crude, it's only a few lines of code to replace it, I would just turn it off. 

You can see the code here:

https://github.com/jcupitt/libvips/blob/master/libvips/iofuncs/image.c#L768

The `if( vips_image_get_typeof( image, "hide-progress" ) ) return;` code is there for nested computations like dzsave, you'll need that in your code too.

**Comment by alonit** [comment]
Hi,

Using `vips_dzsave ` with zip container outputs a single .zip file - Does this .zip file is a part of the Deep Zoom specification? How can I use this .zip file in a deep zoom viewer tool - Do I need to fully extract this .zip file before loading the extracted pyramid into a given deep zoom viewer?

Thanks

**Comment by jcupitt** [comment]
The zip output is to help Windows. It hates large directories with many small files, performance takes a terrible hit, so it's much faster to create a zip file directly. When I timed it last, on Windows zip output was at least 3x faster than filesystem output -- in other words, with FS output, at least 75% of run time is wasted uselessly inside the OS somewhere. 

Plus most people generate pyramids on a different machine than their webserver, so the pyramids need zipping up anyway to copy over the network. And a zipfile uses significantly less space on your disc drive than an equivalent filesystem, since the FS will store one tile per page.

The fs and the zip are the same speed on linux, and zip/unzip is very fast, so there's little advantage there, except perhaps for the space saving. 

For serving, you have several options:

* Unzip the pyramid on the webserver.
* Leave the file zipped, but parse it and extract the index. Write a little PHP/Python/whatever to respond to requests on a URL to pull certain tiles out of the zip on demand, and use mod_rewrite to present your script as a regular URL. Now you can use any viewer without modification and without unzipping.
* Leave the file zipped, but extract the index and place that on the server too. Write your own `tileUrl` Javascript function to download and cache the index file, then use a resume http request to put a tile directly from inside the zip.

**Comment by alonit** [comment]
O.K.  I'll check these options.
Thanks for this detailed and professional answer and for your kind support.

**Comment by jcupitt** [comment]
I think we're done, I'll close. Please open a new issue if you still have problems.
